### PR TITLE
[CLI] introduce local simulation, benchmarking and replay

### PR DESCRIPTION
**Pull request by vgao1996**
## Description
This is a major enhancement to the CLI's ability to perform various kinds of local simulations.

### New Options for `aptos move run/run-script/publish`
These commands now get the following new options:
  - `--local`: simulates the transaction locally as opposed to sending it to the fullnode
  - `--benchmark`: benchmarks the transaction locally

### New Command `aptos move replay`
```
Usage: aptos move replay [OPTIONS] --network <NETWORK> --txn-id <TXN_ID>

Options:
      --network <NETWORK>
          The network to replay on.
          
          Possible values: mainnet, testnet, <REST_ENDPOINT_URL>

      --txn-id <TXN_ID>
          The id of the transaction to replay. Also being referred to as "version" in some contexts

      --benchmark
          If this option is set, benchmark the transaction and report the running time(s)

      --profile-gas
          If this option is set, profile the transaction and generate a detailed report of its gas usage

      --skip-comparison
          If present, skip the comparison against the expected transaction output
```
This offers a convenient way to replay any committed transactions. Benchmarking and gas profiling are also supported.

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [x] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [x] Aptos CLI/SDK
- [x] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?
Right now there isn't an easy way to write these kinds of tests for the CLI, so let's bear with the manual tests for now
```
// Run from `aptos-move/move-examples/hello_blockchain`
cargo run -p aptos -- publish --local --named-addresses hello_blockchain=default
cargo run -p aptos -- publish --benchmark --named-addresses hello_blockchain=default
```
```
cargo run -p aptos -- move replay --network mainnet --txn-id 546684099
cargo run -p aptos -- move replay --network mainnet --txn-id 546684099 --benchmark
cargo run -p aptos -- move replay --network mainnet --txn-id 546684099 --profile-gas
```

## Key Areas to Review
- [crates/aptos/src/common/local_simulation.rs](https://github.com/aptos-labs/aptos-core/pull/12832/files#diff-ede1a113183e6eaf69f0f3212dbf901988b994e19335809ae59c76846ce21d40)
  - Helpers functions to perform local simulation, benchmarking and gas profiling
- [crates/aptos/src/move_tool/mod.rs](https://github.com/aptos-labs/aptos-core/pull/12832/files#diff-ab6f2a30099fea3607dd07fbc396c69729bfe8840558a5c026a1045eb8ac4aba)
  - The new replay command

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: ecosystem/indexer-grpc/indexer-grpc-in-memory-cache-benchmark/Cargo.toml [hunk 1]
```diff
@@ -1,5 +1,6 @@
 [package]
 name = "aptos-indexer-grpc-in-memory-cache-benchmark"
+version = "0.1.0"
```
Comment (runtian-zhou) [inline ecosystem/indexer-grpc/indexer-grpc-in-memory-cache-benchmark/Cargo.toml:3]: Is this relevant change?

File: crates/aptos/src/move_tool/mod.rs [hunk 2]
```diff
@@ -1491,6 +1498,158 @@ impl CliCommand<TransactionSummary> for RunScript {
...
+    fn command_name(&self) -> &'static str {
+        "Replay"
+    }
+
+    async fn execute(self) -> CliTypedResult<TransactionSummary> {
+        use ReplayNetworkSelection::*;
+
+        let rest_endpoint = match &self.network {
+            Mainnet => "https://fullnode.mainnet.aptoslabs.com",
+            Testnet => "https://fullnode.testnet.aptoslabs.com",
```
Comment (runtian-zhou) [inline crates/aptos/src/move_tool/mod.rs:1558]: Add devnet as well?

File: crates/aptos/src/move_tool/mod.rs [hunk 3]
```diff
@@ -1491,6 +1498,158 @@ impl CliCommand<TransactionSummary> for RunScript {
...
+
+        // Fetch the transaction to replay.
+        let (txn, txn_info) = debugger
+            .get_committed_transaction_at_version(self.txn_id)
+            .await?;
+
+        let txn = match txn {
+            Transaction::UserTransaction(txn) => txn,
+            _ => {
+                return Err(CliError::UnexpectedError(
```
Comment (runtian-zhou) [inline crates/aptos/src/move_tool/mod.rs:1575]: This shouldn't stictly needed as the vm should be able to handle system txn as well? But I guess this restriction makes sense as part of the cli?

File: crates/aptos/src/move_tool/mod.rs [hunk 4]
```diff
@@ -1491,6 +1498,158 @@ impl CliCommand<TransactionSummary> for RunScript {
...
+        // Execute the transaction.
+        let (vm_status, vm_output) = if self.profile_gas {
+            println!("Profiling transaction...");
+            local_simulation::profile_transaction_using_debugger(
+                &debugger,
+                self.txn_id,
+                txn.clone(),
+                hash,
+            )?
+        } else if self.benchmark {
```
Comment (runtian-zhou) [inline crates/aptos/src/move_tool/mod.rs:1593]: Should it be `if` or `else if` here? Seems that you want to run it one by one instead of only selecting one execution mode?

**Review by runtian-zhou — APPROVED** [review]
Beautiful. Can you add example output in the PR description as well?

File: ecosystem/indexer-grpc/indexer-grpc-in-memory-cache-benchmark/Cargo.toml (hunk 1 above)
Comment (vgao1996) [reply ecosystem/indexer-grpc/indexer-grpc-in-memory-cache-benchmark/Cargo.toml:3]: Well what I found out was that `cargo fix` would complain about this not having a version but strangely `cargo build` does not. Decided to fix it as `cargo fix` can be a really helpful command for removing unused imports.

File: crates/aptos/src/move_tool/mod.rs (hunk 3 above)
Comment (vgao1996) [reply crates/aptos/src/move_tool/mod.rs:1575]: Good catch. It's more of a limitation of the current VM APIs -- the ones suitable for this use case currently expect a `SignedTransaction` for reasons I don't fully understand.

It should be possible for us to extend support to other kinds of transactions, but that'll likely require some more refactoring of those APIs themselves. Let me leave a comment here and mention this as a future improvement.

File: crates/aptos/src/move_tool/mod.rs (hunk 4 above)
Comment (vgao1996) [reply crates/aptos/src/move_tool/mod.rs:1593]: To keep things simple I should probably just make these options mutually exclusive. Will update the PR to reflect this.

Running it one by one may work for benchmark and profile-gas, but once you add another option "--step", things may start to look ugly.

What's your preference?

Comment (runtian-zhou) [reply crates/aptos/src/move_tool/mod.rs:1593]: Yep mutually exclusive also makes sense to me.

File: aptos-move/aptos-debugger/src/aptos_debugger.rs [hunk 5]
```diff
@@ -133,7 +134,7 @@ impl AptosDebugger {
 
                     // Deprecated.
                     TransactionPayload::ModuleBundle(..) => {
-                        unreachable!("Module bundle payload has already been checked")
+                        unreachable!("Module bundle payload has already been deprecated")
```
Comment (georgemitenkov) [inline aptos-move/aptos-debugger/src/aptos_debugger.rs:137]: This messages said checked because before this function is called, we already reject module bundles. So it is truly unreachable code because of the check

File: aptos-move/aptos-debugger/src/aptos_debugger.rs [hunk 6]
```diff
@@ -98,7 +98,8 @@ impl AptosDebugger {
             .map_err(|err| format_err!("Unexpected VM Error: {:?}", err))?;
 
         // TODO(Gas): revisit this.
-        let resolver = state_view.as_move_resolver();
+        let resolver: aptos_vm::data_cache::StorageAdapter<'_, DebuggerStateView> =
```
Comment (georgemitenkov) [inline aptos-move/aptos-debugger/src/aptos_debugger.rs:101]: Why this change?

**Review by georgemitenkov — APPROVED** [review]
Amazing! We should probably be careful with using the replay results as is because we run a single transaction without block context (so some things like block gas limit, aggregators are not observable). But these can be added later

File: crates/aptos/src/move_tool/mod.rs [hunk 7]
```diff
@@ -1491,6 +1498,158 @@ impl CliCommand<TransactionSummary> for RunScript {
...
+        let resolver = state_view.as_move_resolver();
+
+        let txn_output = vm_output
+            .try_materialize_into_transaction_output(&resolver)
+            .map_err(|err| {
+                CliError::UnexpectedError(format!(
+                    "Failed to materialize into transaction output: {}",
+                    err
+                ))
+            })?;
```
Comment (vgao1996) [inline crates/aptos/src/move_tool/mod.rs:1622]: @georgemitenkov good point on aggregators. However since I do try to materialize the vm output here, I think aggregators are observable, no?

Regarding the block gas limit, I don't think it matters for replay since that is only used to terminate a pending block early, and committed are transactions don't really have the concept of a block any more. @igor-aptos please correct me if I'm wrong on this.

File: aptos-move/aptos-debugger/src/aptos_debugger.rs (hunk 5 above)
Comment (vgao1996) [reply aptos-move/aptos-debugger/src/aptos_debugger.rs:137]: I see I didn't realize there's a check before this. Will revert it back but maybe add a bit more context

File: crates/aptos/src/move_tool/mod.rs (hunk 7 above)
Comment (georgemitenkov) [reply crates/aptos/src/move_tool/mod.rs:1622]: V1 - yes. New aggregators only exist if we execute a block, so their side-effects are not materialized here. I suspect that it is okay though because VM is created as `let vm = AptosVM::new(&resolver, None);` where because of None we ensure that aggregators are off in this context.

But the best way to know is to simulate a testnet txn with aggregator V2 (I can check that after)

File: aptos-move/aptos-debugger/src/aptos_debugger.rs (hunk 6 above)
Comment (vgao1996) [reply aptos-move/aptos-debugger/src/aptos_debugger.rs:101]: Hmm... Not sure why I changed this... Could be the legacy of some rebasing. Got it reverted.
